### ISSUE TITLE
Arbitrary Pulumi commands

### ISSUE THREAD

**Issue by JimMadge** [alan-turing-institute/data-safe-haven#1715]
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a request for a new feature in the Data Safe Haven or an upgrade to an existing feature.
- [x] The feature is still missing in the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :strawberry: Suggested change

A few times in conversation the idea has come up of how to run arbitrary pulumi commands using the stacks and pulumi configuration (`pulumi.yaml`, `pulumi.<stack>.yaml`) that the code manages. @jemrobinson I think this came up when we spoke last.

## :steam_locomotive: How could this be done?

I think it should be fairly simple to programmatically load the appropriate stacks and configuration before passing arguments to the pulumi CLI.
There is some risk there of ACE.
I would expect this _isn't_ something we would advice to be used to normal operations. I do think it could be very useful for development/testing/debugging though.

**Comment by jemrobinson** [comment]
To run on the command line you need to know the storage account name and storage key. You can then run commands from the directory that has the `Pulumi.<stack name>.yaml` file as follows:

```console
AZURE_STORAGE_ACCOUNT='<storage account name>' AZURE_STORAGE_KEY='<key here>' AZURE_KEYVAULT_AUTH_VIA_CLI='true' pulumi refresh -p 1
```

**Comment by JimMadge** [comment]
Yes, that's right. We have the routines to get that information (from context settings) already.

I still didn't get to look at whether this all works as an "inline program" so there is no need for local Pulumi files.

**Comment by JimMadge** [comment]
@jemrobinson @craddm Just noting here that something like this seems to be useful for developers, for example,

- fixing problems
- `pulumi import`

**Comment by JimMadge** [comment]
Closed in #1888
